testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java

```diff
@@ -13824,6 +13824,54 @@ void testTimestampDiff(boolean coercionEnabled) {
     f.checkNull("date_add(CAST(NULL AS DATE), interval 5 day)");
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-6527">[CALCITE-6527]
+   * Add DATE_ADD function (enabled in Spark library)</a>.
+   */
+  @Test void testDateAddSpark() {
+    final SqlOperatorFixture f0 = fixture()
+        .setFor(SqlLibraryOperators.DATE_ADD_SPARK);
+    f0.checkFails("^date_add(date '2008-12-25', "
+            + "5)^",
+        "No match found for function signature "
+            + "DATE_ADD\\(<DATE>, <NUMERIC>\\)", false);
+
+    final SqlOperatorFixture f = f0.withLibrary(SqlLibrary.SPARK);
+    f.checkScalar("date_add(date '2016-02-22', 2)",
+        "2016-02-24",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(date '2016-02-28', 2)",
+        "2016-03-01",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(date '2016-03-01', -2)",
+        "2016-02-28",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(date '2016-03-01', -2)",
+        "2016-02-28",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(date '2016-03-01', -2.0)",
+        "2016-02-28",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(date '2016-02-28', 2.0)",
+        "2016-03-01",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(date '2016-02-28', '2.0')",
+        "2016-03-01",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(timestamp '2016-02-22 13:00:01', '-2.0')",
+        "2016-02-20",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(timestamp '2016-02-22 13:00:01', -2)",
+        "2016-02-20",
+        "DATE NOT NULL");
+    f.checkScalar("date_add(timestamp '2016-02-22 13:00:01', -2.0)",
+        "2016-02-20",
+        "DATE NOT NULL");
+    f.checkNull("date_add(CAST(NULL AS DATE), 5)");
+    f.checkNull("date_add(date '2016-02-22', CAST(NULL AS INTEGER))");
```

Comment:
how about a few tests around year 0 and the Gregorian calendar cutoff date?
For example, date_add('0001-01-01', -2). Notice that there is no year 0, so this probably should throw.

Reply:
Good suggestion. I tested it in spark and it seems that spark is not aware of this error. Should we fix this error in calcite or keep this behavior?
scala> val df = spark.sql("SELECT date_add('0001-01-01', -2)")
df: org.apache.spark.sql.DataFrame = [date_add(0001-01-01, -2): date]

```
scala> df.show()
+------------------------+
|date_add(0001-01-01, -2)|
+------------------------+
|              0000-12-30|
+------------------------+


scala> val df = spark.sql("SELECT date_add('0000-01-01', -2)")
df: org.apache.spark.sql.DataFrame = [date_add(0000-01-01, -2): date]

scala> df.show()
+------------------------+
|date_add(0000-01-01, -2)|
+------------------------+
|             -0001-12-30|
+------------------------+

```

Reply:
You should check the issues open for spark, this looks like a bug to me. There is no year 0.
If there is no bug maybe you can open one.

Reply:
OK, I will try again in the latest spark code. If there is no corresponding jira record, I will file a jira case to describe this bug.

I feel that 000-00-01 is reasonable, but negative numbers are unreasonable. Calcite currently follows this principle. The following are my test results in Calcite.
```
    f.checkScalar("date_add('0001-01-01', -2)",
        "0000-12-30",
        "DATE NOT NULL");
    f.checkFails("date_add('0000-01-01', -2)",
        "Value of YEAR field is out of range in '0000-01-01'",
        true);

```

Reply:
on the contrary, year 0 is not legal, it never existed.
and people lived before Christ, so years BC are completely legal in many databases. whether they show up as negative numbers, that's a different question. Calcite doesn't support them, though

Reply:
I have already raised a jira case in spark
https://issues.apache.org/jira/browse/SPARK-49247

Reply:
I think that in my previous review I asked for a few more tests around boundary conditions, I don't see any here.

Reply:
_Au contraire_. `0000-00-00` is a valid 8601 date.

See https://en.wikipedia.org/wiki/ISO_8601#Calendar_dates

Reply:
Will the Calcite DATE datatype accept such a value?

Reply:
I think I will do some tests and describe them later. It seems that calcite has done some processing but it is not perfect. I will fix it in this PR

Reply:
@mihaibudiu, I don't recall. I just happened to reading about ISO-8601 so I thought I'd correct misinformation.

Reply:
Sorry for the late reply.
After testing, the date_add function will format the date. For example, invalid dates such as `0000-01-01` will fail the date format.